### PR TITLE
Beanstream Gateway Improvements

### PR DESCRIPTION
**Pull request by lumiit-adam**
Integrated the IPP gateway into the Beanstream gatewayand use BeanstreamGateway as the unified entrance;

remove IPP gateway files;

Because the two gateways belong to the same company just different branch. So the developers from the company hope to make them an unified gateway to Shopify.

(Beansteam website: http://www.beanstream.com/)

### CODE REVIEW

**Comment by elfassy** [comment]
looks good to me, please squash into a single commit

**Comment by WizardCloud** [comment]
@elfassy squashed now

**Comment by elfassy** [comment]
@lumiit-adam can you post a link to the docs in the description, thx

File: lib/active_merchant/billing/gateways/beanstream/ipp_core.rb [hunk 1]
```diff
@@ -151,23 +160,23 @@ def new_submit_xml(action)
         xml.target!
       end
 
-      def commit_url
-        (test? ? test_url : live_url)
+      def ipp_commit_url
+        (test? ? IPP_TEST_URL : IPP_LIVE_URL)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/ipp_core.rb]: please split the commits into two 
1) renaming (`_ipp`, `remote_beanstream_ipp_test.rb`...)
2) code changes
thx

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 2]
```diff
@@ -64,31 +65,128 @@ module Billing #:nodoc:
...
+      # capture or refund transactions then you'll also need to add a username
+      # and password to your account under administration -> account settings ->
+      # order settings -> Use username/password validation
+      def initialize(options = {})
+        #region: 0 - US & CA, 1 - AU & NZ
+        @region = empty?(options[:region]) ? 0 : options[:region]
+
+        # I want to add this logic to judge the region from shipping & billing address
+        # but don't know if it is a correct way.
+        # I think this can be decided by shopify which will integrate with.
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: @lumiit-adam if I understand correctly, `options[:region]` is a new option what would need to be passed in, and wasn't used previouly?

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 3]
```diff
@@ -191,19 +289,15 @@ def update(vault_id, payment_method, options = {})
...
       end
 
-      def scrub(transcript)
-        transcript.
-          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
-          gsub(/(&?password=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
+      def include_beanstream
+        BeanstreamGateway.include BeanstreamCore
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: I don't think this will work as you expect it to. The module will be included on the class, not on only for a specific instance.
Consider using a factory pattern (https://stackoverflow.com/a/41642955/454375)

Comment (elfassy) [reply lib/active_merchant/billing/gateways/beanstream.rb]: could look something like
```ruby
class BeanstreamGateway < Gateway
  def self.new(options = {})
    region = get_region_from_options(options)
    (region == 1 ? BeanstreamIppGateway : BeanstreamDefautGateway).new(options)
  end
end

class BeanstreamIppGateway < Gateway
  ...
end

class BeanstreamDefautGateway < Gateway
  include BeanstreamCore
  ...
end
```

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 2 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: If there is other way can pass the parameter without add new option, please tell me.

File: lib/active_merchant/billing/gateways/beanstream/ipp_core.rb [hunk 4]
```diff
@@ -2,28 +2,31 @@
 
 module ActiveMerchant #:nodoc:
   module Billing #:nodoc:
-    class IppGateway < Gateway
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/ipp_core.rb]: please revert the changes to this file. The `IppGateway` might be used and cannot be changed just for beanstream. Instead bring the variations to `BeanstreamIppGateway`.

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 5]
```diff
@@ -64,7 +65,54 @@ module Billing #:nodoc:
...
+        #   end
+        # end
+
+        region = options[:region]
+
+        (region == 1 ? BeanstreamIppGateway : BeanstreamDefautGateway).new(options)
+      end
+    end
+
+    class BeanstreamIppGateway < Gateway
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: please move each class to it's own file

**Comment by lumiit-adam** [comment]
I have changed code, please continue to review.

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 6]
```diff
@@ -64,148 +65,46 @@ module Billing #:nodoc:
...
+      def initialize(options = {})
+        @region = empty?(options[:region]) ? 0 : options[:region]
 
-        if payment_method.respond_to?(:number)
-          add_credit_card(post, payment_method)
+        if @region == 1
+          requires!(options, :username, :password, :login)
         else
-          post[:singleUseToken] = payment_method
+          requires!(options, :login)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: can move `requires!` to correct class

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 7]
```diff
@@ -78,13 +78,7 @@ class BeanstreamGateway < Gateway
...
 
       def initialize(options = {})
-        @region = empty?(options[:region]) ? 0 : options[:region]
-
-        if @region == 1
-          requires!(options, :username, :password, :login)
-        else
-          requires!(options, :login)
-        end
+        requires!(options, :region)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: this needs to be be backward compatible, make region 0 the default

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 8]
```diff
@@ -64,148 +65,40 @@ module Billing #:nodoc:
...
-        end
-        add_secure_profile_variables(post, options)
+    class BeanstreamGateway < Gateway
+      attr_accessor :region
+      include Empty
 
-        commit(post, true)
-      end
+      def initialize(options = {})
+        requires!(options, :region)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: no need to require since it's it nil it will default to the old behaviour

File: lib/active_merchant/billing/gateways/beanstream/ipp_core.rb [hunk 9]
```diff
@@ -0,0 +1,184 @@
+require "nokogiri"
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    module IPPCore
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/ipp_core.rb]: don't reopen this class, move over the differences to `BeanstreamIppGateway` and remove this file

File: lib/active_merchant/billing/gateways/beanstream/beanstream_ipp.rb [hunk 10]
```diff
@@ -0,0 +1,179 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class BeanstreamIppGateway < Gateway
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/beanstream_ipp.rb]: is it possible to inherit from `IppGateway`?

File: test/fixtures.yml [hunk 11]
```diff
@@ -393,8 +393,9 @@ instapay:
   login: TEST0
   password:
 
-# Working credentials, no need to replace
+# login: merchant id
```
Comment (elfassy) [inline test/fixtures.yml]: why the change in comment?

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 12]
```diff
@@ -64,148 +65,34 @@ module Billing #:nodoc:
...
+        # but don't know if it is a correct way.
+        # I think this can be decided by shopify which will integrate with.
 
-      def supports_scrubbing?
-        true
-      end
+        # [ options[:billing_address], options[:shipping_address] ].compact.each do |address|
+        #   next if empty?(address[:country])
+        #   unless ['AU', 'NZ'].include?(address[:country])
+        #     @region = 1
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: when is IPP required? Does it have to do with the shipping and/or billing address?

File: test/fixtures.yml (hunk 11 above)
Comment (lumiit-adam) [reply test/fixtures.yml]: I think there is no exact function for this change, I will revert it.

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 12 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: I am not clear about tht. The staff from ipp just asked for the unify gateway for different regions' orders. So, I think that if there is a way to judge the region automatically? Could you give me some suggestion about it?

File: lib/active_merchant/billing/gateways/beanstream/beanstream_ipp.rb (hunk 10 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream/beanstream_ipp.rb]: I think I can do it.

**Comment by chriswongipp** [comment]
@elfassy is there anything outstanding before this can be merged into Master?

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 13]
```diff
@@ -65,147 +66,25 @@ module Billing #:nodoc:
...
-      end
-
-      def scrub(transcript)
-        transcript.
-          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
-          gsub(/(&?password=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
-      end
+        region = options[:region]
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: what does the region represent in this context? Would it not make more sense to keep the gateways split?

File: CHANGELOG [hunk 14]
```diff
@@ -138,6 +138,7 @@
 * Spreedly: Scrub sensitive transaction data [abarrak] #2781
 * Stripe: Add `exchange_rate` parameter [WilsonChiang] #2796
 * Mundipagg: New Gateway Implementation [nfarve] #2791
+* Integrated the Bambora Pacifc gateway into Bambora North America gateway #2797
```
Comment (elfassy) [inline CHANGELOG]: please put the name of the gateway first followed by `:`

File: test/remote/gateways/remote_ipp_test.rb [hunk 15]
```diff
@@ -81,4 +81,4 @@ def test_invalid_login
     response = gateway.purchase(200, @credit_card, @options)
     assert_failure response
   end
-end
+end
```
Comment (elfassy) [inline test/remote/gateways/remote_ipp_test.rb]: ✂️

File: CHANGELOG (hunk 14 above)
Comment (lumiit-adam) [reply CHANGELOG]: Fixed.

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 16]
```diff
@@ -65,147 +66,25 @@ module Billing #:nodoc:
...
+        # I think this can be decided by shopify which will integrate with.
 
-      def cancel_recurring(options = {})
-        ActiveMerchant.deprecated RECURRING_DEPRECATION_MESSAGE
+        # [ options[:billing_address], options[:shipping_address] ].compact.each do |address|
+        #   next if empty?(address[:country])
+        #   unless ['AU', 'NZ'].include?(address[:country])
+        #     @region = 1
+        #   end
+        # end
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: ✂️ please remove these comments. To make the default explicit, please modify the following line to be 
```
region = options[:region] || 0
```

File: lib/active_merchant/billing/gateways/ipp.rb [hunk 17]
```diff
@@ -172,4 +172,4 @@ def authorization_from(response)
       end
     end
   end
-end
+end
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/ipp.rb]: ✂️

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 16 above)
Comment (elfassy) [reply lib/active_merchant/billing/gateways/beanstream.rb]: an alternative to integers would be to have symbols represent the regions, example: `:pacific`, `:america`... This will make it easier to know what region is being used without having to refer back to this file

File: test/unit/gateways/ipp_test.rb [hunk 18]
```diff
@@ -208,4 +208,4 @@ def successful_refund_response
 </SubmitSingleRefundResult></SubmitSingleRefundResponse></soap:Body></soap:Envelope>
     XML
   end
-end
+end
```
Comment (elfassy) [inline test/unit/gateways/ipp_test.rb]: ✂️

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 19]
```diff
@@ -65,147 +66,25 @@ module Billing #:nodoc:
     #     :custom => 'reference one'
     #   )
     class BeanstreamGateway < Gateway
-      include BeanstreamCore
+      attr_accessor :region
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: ✂️

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 20]
```diff
@@ -65,147 +66,25 @@ module Billing #:nodoc:
...
-          gsub(/(&?password=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
-      end
+        region = options[:region]
 
-      private
-      def build_response(*args)
-        Response.new(*args)
+        (region == 1 ? BeanstreamIppGateway : BeanstreamDefautGateway).new(options)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]:
```ruby
klass = 
  case options[:region]
  when nil, :america
    BeanstreamDefautGateway
  when :pacific
     BeanstreamIppGateway
  else
    raise ArgumentError, "invalid region #{region}"
  end
klass.new(options)
```

**Comment by lumiit-adam** [comment]
Finished squashing commit and waiting for final reviewing

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 21]
```diff
@@ -64,149 +65,29 @@ module Billing #:nodoc:
...
-          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
-          gsub(/(&?password=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?passcode=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
-      end
+    #
+    # Modification on 2018-03-12:
+    # New feature: integrated the ipp gateway into beanstream gateway, use this class as an unified entrance for both the gateways
+    # Get the :region field from options parameter, :region: 0 - US & CA, 1 - AU
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: comment no longer reflects the code, symbol should be passed instead of integers

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 22]
```diff
@@ -64,149 +65,29 @@ module Billing #:nodoc:
...
+    # Proess payment request from US & CA using beanstream service (default)
+    # Proess payment request from AU using ipp service
+    # Will remove ipp gateway file
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream.rb]: `Will remove ipp gateway file` not sure what you mean by this comment

File: lib/active_merchant/billing/gateways/beanstream/beanstream_core.rb [hunk 23]
```diff
@@ -142,16 +142,7 @@ def self.included(base)
...
         base.display_name = 'Beanstream.com'
-      end
-
-      # Only <tt>:login</tt> is required by default,
-      # which is the merchant's merchant ID. If you'd like to perform void,
-      # capture or refund transactions then you'll also need to add a username
-      # and password to your account under administration -> account settings ->
-      # order settings -> Use username/password validation
-      def initialize(options = {})
-        requires!(options, :login)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/beanstream_core.rb]: why was this removed?

File: lib/active_merchant/billing/gateways/beanstream/beanstream_default.rb [hunk 24]
```diff
@@ -0,0 +1,178 @@
+require 'active_merchant/billing/gateways/beanstream/beanstream_core'
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class BeanstreamDefautGateway < Gateway
+      include BeanstreamCore
+
+      def initialize(options = {})
+        requires!(options, :login)
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/beanstream_default.rb]: let's leave this in BeanstreamCore to avoid modifying that file, since it could be used elsewhere

File: lib/active_merchant/billing/gateways/beanstream/ipp_core.rb [hunk 25]
```diff
@@ -0,0 +1,183 @@
+require 'nokogiri'
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    module IPPCore
```
Comment (elfassy) [inline lib/active_merchant/billing/gateways/beanstream/ipp_core.rb]: where is this used?

File: test/unit/gateways/beanstream_ipp_test.rb [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+-> "Connection: close\r\n"
+-> "\r\n"
+reading 767 bytes...
+-> "<?xml version=\"1.0\" encoding=\"utf-8\"?><soap:Envelope xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\"><soap:Body><SubmitSinglePaymentResponse xmlns=\"http://www.ippayments.com.au/interface/api/dts\"><SubmitSinglePaymentResult>&lt;Response&gt;\r\n\t&lt;ResponseCode&gt;1&lt;/ResponseCode&gt;\r\n\t&lt;Timestamp&gt;20-Dec-2014 06:55:17&lt;/Timestamp&gt;\r\n\t&lt;Receipt&gt;&lt;/Receipt&gt;\r\n\t&lt;SettlementDate&gt;&lt;/SettlementDate&gt;\r\n\t&lt;DeclinedCode&gt;183&lt;/DeclinedCode&gt;\r\n\t&lt;DeclinedMessage&gt;Exception parsing transaction XML&lt;/DeclinedMessage&gt;\r\n&lt;/Response&gt;\r\n</SubmitSinglePaymentResult></SubmitSinglePaymentResponse></soap:Body></soap:Envelope>"
+read 767 bytes
+Conn close
+    PRE_SCRUBBED
+  end
+
+  def post_scrubbed
```
Comment (elfassy) [inline test/unit/gateways/beanstream_ipp_test.rb]: could we use the VCR gem to record these responses? It would be easier to re-run in the future

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 21 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: Fixed.

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 22 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: Fixed.

File: lib/active_merchant/billing/gateways/beanstream/beanstream_core.rb (hunk 23 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream/beanstream_core.rb]: Fixed.

File: lib/active_merchant/billing/gateways/beanstream/beanstream_default.rb (hunk 24 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream/beanstream_default.rb]: Fixed.

File: lib/active_merchant/billing/gateways/beanstream/ipp_core.rb (hunk 25 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream/ipp_core.rb]: Deleted.

File: test/unit/gateways/beanstream_ipp_test.rb (hunk 26 above)
Comment (lumiit-adam) [reply test/unit/gateways/beanstream_ipp_test.rb]: If we can leave this feature in future improvement?

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 27]
```diff
@@ -64,149 +65,28 @@ module Billing #:nodoc:
...
-          gsub(/(&?passcode=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
-      end
+    #
+    # Modification on 2018-03-12:
+    # New feature: integrated the ipp gateway into beanstream gateway, use this class as an unified entrance for both the gateways
+    # Get the :region field from options parameter, :america (nil) - US & CA, :pacific - AU & NZ
+    # Proess payment request from US & CA using beanstream service (default)
+    # Proess payment request from AU & NZ using ipp service
```
Comment (pi3r) [inline lib/active_merchant/billing/gateways/beanstream.rb]: The modifications shouldn't be stored as comments, it should be in the commit message body

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 28]
```diff
@@ -64,149 +65,28 @@ module Billing #:nodoc:
...
 
-      private
-      def build_response(*args)
-        Response.new(*args)
+    class BeanstreamGateway < Gateway
+      include Empty
```
Comment (pi3r) [inline lib/active_merchant/billing/gateways/beanstream.rb]: Why is `Empty` included here? I can't find where it's used.

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 29]
```diff
@@ -64,149 +65,28 @@ module Billing #:nodoc:
...
+    # Modification on 2018-03-12:
+    # New feature: integrated the ipp gateway into beanstream gateway, use this class as an unified entrance for both the gateways
+    # Get the :region field from options parameter, :america (nil) - US & CA, :pacific - AU & NZ
+    # Proess payment request from US & CA using beanstream service (default)
+    # Proess payment request from AU & NZ using ipp service
 
-      private
-      def build_response(*args)
-        Response.new(*args)
+    class BeanstreamGateway < Gateway
```
Comment (pi3r) [inline lib/active_merchant/billing/gateways/beanstream.rb]: What's the point of inheriting from `Gateway` if it's never meant to be used as a gateway? Here `new` is a factory method that will return a `Gateway` object but `BeanstreamGateway` is not meant to be one.

Comment (pi3r) [reply lib/active_merchant/billing/gateways/beanstream.rb]: @bdewater Can I have some thoughts on that? Wondering if anything will be used at the class level and `Beanstream` does need to impl `Gateway` even if it doesn't use it.

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 27 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: Finish this change request.

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 28 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: Finish this change request.

File: lib/active_merchant/billing/gateways/beanstream.rb (hunk 29 above)
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: Finish this change request.

Comment (pi3r) [reply lib/active_merchant/billing/gateways/beanstream.rb]: @curiousepic pinging you here because I'd like a second opinion on that. Hijacking `.new` to implement a factory feels wrong to me but I'm not sure if you have any experience on how to do that in AM. Also, should the "factory gateway" inherit from `Gateway`?

File: CHANGELOG [hunk 30]
```diff
@@ -152,6 +152,7 @@
 * Spreedly: Scrub sensitive transaction data [abarrak] #2781
 * Stripe: Add `exchange_rate` parameter [WilsonChiang] #2796
 * Mundipagg: New Gateway Implementation [nfarve] #2791
+* Beanstream: Integrated the Bambora Pacifc gateway into Bambora North America gateway #2813
```
Comment (bdewater) [inline CHANGELOG]: Can you make sure this changelog line is moved up? This version (1.78.0) was released a while ago.

File: lib/active_merchant/billing/gateways/beanstream.rb [hunk 31]
```diff
@@ -64,149 +65,20 @@ module Billing #:nodoc:
...
-          gsub(/(&?password=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?passcode=)[^&\s]*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardCvd=)\d*(&?)/, '\1[FILTERED]\2').
-          gsub(/(&?trnCardNumber=)\d*(&?)/, '\1[FILTERED]\2')
-      end
-
-      private
-      def build_response(*args)
-        Response.new(*args)
+    class BeanstreamGateway
```
[the remark this answers is not in the document]
Comment (elfassy) [reply lib/active_merchant/billing/gateways/beanstream.rb]: oups, same comment as @pi3r :D

[the remark this answers is not in the document]
Comment (lumiit-adam) [reply lib/active_merchant/billing/gateways/beanstream.rb]: I update the commits to let the BeanstreamGateway to inherit from Gateway. Please have a review.

**Comment by elfassy** [comment]
Any reason we must use the region option? Seems to me like it would be best to let the end user decide between `BeanstreamDefautGateway` and `BeanstreamIppGateway` . Each already provides the information as to where it can be used (https://github.com/activemerchant/active_merchant/pull/2813/files#diff-1db73f09eabc8b8a2be54b728f148ee9R7)
